p2p/proto/block.proto

```diff
@@ -72,10 +80,12 @@ message BlockBodiesResponse {
     optional BlockID id = 1; // may not appear if Fin is sent to end the whole response
 
     oneof body_message {
-        StateDiff  diff    = 2;
-        Classes    classes = 3;
-        BlockProof proof   = 4;
-        Fin        fin     = 5;
+        StateDiff    diff         = 2;
+        Classes      classes      = 3;
+        BlockProof   proof        = 4;
+        Transactions transactions = 6;
```

Comment:
transactoins and events were not supposed to be part of the block body. this can be discussed. the idea was that with stark proofs we don't need to validate transactions by executing them. This is when there's proof which is K blocks behind. Until then the idea was to rely on the consensus (similarly to how Ethereum has block times and epochs). 
The protocol can be defined so that txs are mandatory until K and then are optional. 
Similarly for receipts & events (which were separated).

minor comment: numbering jumps from 4 to 6

Reply:
Are all forms of nodes expected to serve the RPC spec? If so, then having transactions and receipts as optional until after K blocks seems unnecessary because all nodes would be fetching them separately to serve the RPC requests. And since block bodies are chunked I don't see any benefit in fetching them separately as that would increase network overhead.

> minor comment: numbering jumps from 4 to 6

I will update

Reply:
It means that to have the data needed to sync with the chain, a node doesn't need to keep transactions before the block with the proof. They can clean their DB. If a node wants to, it can of course keep whatever it wants. 
As mentioned, they are not part of the block body, so no need for a client to fetch them. Even if they are included for new blocks, they would be optional and with a separate endpoint to fetch them (from those nodes that keep them beyond the optional point)

Reply:
What defines whether something should be a part of the block body?

We don't need `State Diff` or `Classes` to be part of the block body because State Diff has a commitment which includes the Classes declared in the blocks, therefore, a peer can fetch them separately and verify the state diff commitment with the signature over the block hash and state diff commitment. Note we will have to add `Classes` to `State Diff` and State diff commitment over the state diff defined by feeder gateway not p2p state diff.

Why can't we include the BlockProof in the BlockHeader? Given other fields are so small `142KB+ other fields' size < 1MB`.

If we make the above changes the block bodies can be removed.

Reply:
Right, they can fetch them separately. So the thought was what will peers have to fetch in order to be an operational node. They need the current state, they don't need past transactions. 

I agree the proof may be part of the header, as optional. Note that we can think of returning many headers in one message (and then it makes sense to have the proof outside)

Reply:
> So the thought was what will peers have to fetch in order to be an operational node.

Can you please elaborate on what you mean by an operational node and fetch which messages? 

Also, fetching can be done in order but there is no guarantee the messages will arrive in order. So I am not sure what you mean.

Reply:
I was replying to "... includes the Classes declared in the blocks, therefore, a peer can fetch them separately". Operational means a node that can follow the chain. Such a node doesn't need transactions. If a client wants to check transactions state, they will enable that in a node, if they just want to track events, they will enable that. If they want to check state, they can do that. 

What messages won't arrive in order? If we support out of order messages in a stream then each will have a reference to be able to order them, if needed.

Reply:
> Operational means a node that can follow the chain. Such a node doesn't need transactions. If a client wants to check transactions state, they will enable that in a node, if they just want to track events, they will enable that. If they want to check state, they can do that.

Why is the RPC spec not sufficient to serve this? I don't see the need to replicate this behaviour on the p2p side. Also, how does a peer verify that the events (or any other date) received are correct?

> What messages won't arrive in order? If we support out of order messages in a stream then each will have a reference to be able to order them, if needed.

Messages from a single peer will indeed come in order over a stream but libp2p doesn't allow for ordered messages from multiple peers on a single stream without further multiplexing. If a peer requests information for a block from multiple peers that peer will have multiple streams open, per peer, and thus there is no guarantee block messages part will arrive in order.